pkg/auth/remote_authorizer.go

```diff
@@ -0,0 +1,187 @@
+	}
+	// Hash the request to use as a cache key to save memory.
+	requestKey := sha256.Sum256(requestBytes)
+
+	for {
+		a.lock.Lock()
+		now := a.clock.Now()
+		if response, ok := a.getAndTouchCacheEntry(requestKey); ok && response.expirationTime.After(now) {
+			a.lock.Unlock()
+			return response.err
+		}
+		// No valid cache entry available. Deduplicate requests.
+		if responseReady, ok := a.pendingRequests[requestKey]; ok {
+			a.lock.Unlock()
+			// Wait for the remote request to finish.
+			select {
+			case <-ctx.Done():
+				return util.StatusFromContext(ctx)
+			case <-responseReady:
+				a.lock.Lock()
+				// Check whether the remote authentication call succeeded.
+				response, ok := a.cachedResponses[requestKey]
+				a.lock.Unlock()
+				if ok {
+					// Note that the expiration time is not checked, as the
+					// response is as fresh as it can be.
+					return response.err
+				}
+				// The remote authentication call failed. Retry.
+			}
+		} else {
+			// No pending request. Create one.
+			responseReady := make(chan struct{})
+			a.pendingRequests[requestKey] = responseReady
+			a.lock.Unlock()
+			// Perform the remote authentication request.
+			response, err := a.authorizeRemotely(ctx, request)
+			a.lock.Lock()
+			delete(a.pendingRequests, requestKey)
+			close(responseReady)
+			if err != nil {
+				a.lock.Unlock()
+				return err
+			}
+			a.insertCacheEntry(requestKey, response)
+			a.lock.Unlock()
+			return response.err
+		}
+	}
+}
+
+func (a *remoteAuthorizer) getAndTouchCacheEntry(requestKey RemoteAuthorizerCacheKey) (remoteAuthorizerResponse, bool) {
+	if entry, ok := a.cachedResponses[requestKey]; ok {
+		// Cache contains a matching entry.
+		a.evictionSet.Touch(requestKey)
+		return entry, true
+	}
+	return remoteAuthorizerResponse{}, false
+}
+
+func (a *remoteAuthorizer) insertCacheEntry(requestKey RemoteAuthorizerCacheKey, response remoteAuthorizerResponse) {
+	if _, ok := a.cachedResponses[requestKey]; ok {
+		a.evictionSet.Touch(requestKey)
+	} else {
+		// Cache contains no matching entry. Free up space, so that the
+		// caller may insert a new entry.
+		for len(a.cachedResponses) >= a.maximumCacheSize {
+			delete(a.cachedResponses, a.evictionSet.Peek())
+			a.evictionSet.Remove()
+		}
+		a.evictionSet.Insert(requestKey)
+	}
+	a.cachedResponses[requestKey] = response
+}
+
+func (a *remoteAuthorizer) authorizeRemotely(ctx context.Context, request *auth_pb.AuthorizeRequest) (remoteAuthorizerResponse, error) {
+	ret := remoteAuthorizerResponse{
+		// The default expirationTime has already passed.
+		expirationTime: time.Time{},
+	}
+
+	response, err := a.remoteAuthClient.Authorize(ctx, request)
+	if err != nil {
+		return ret, util.StatusWrapWithCode(err, codes.PermissionDenied, "Remote authorization failed")
+	}
+
+	// An invalid expiration time indicates that the response should not be cached.
+	if response.GetCacheExpirationTime().IsValid() {
```

Comment:
Please only use `.Get*()` in case the containing message may be nil. In this specific case `response` is guaranteed to be non-nil, so use `.CacheExpirationTime` instead.

Also instead of using `.IsValid()`, I'd recommend comparing against `nil` and having a separate `.CheckValid()` to return an error. The downside of the current approach is that in case some incompatibility arises in the way timestamps are sent over, the client will silently discard it.